File: 04-verbose-hello-world-srv/00-verbose-hello-world-srv-basic/src/main/zava/com/github/jinahya/hello/misc/c02rfc862/Rfc862Tcp1Client.py
#!/usr/bin/python3

###
# #%L
# verbose-hello-world-srv-common
# %%
# Copyright (C) 2018 - 2023 Jinahya, Inc.
# %%
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#      http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# #L%
###

import hashlib
import random
import socket
import sys
from base64 import b64encode
from contextlib import closing

IPv4 = True
FAMILY = socket.AF_INET if IPv4 else socket.AF_INET6
HOST = '127.0.0.1' if IPv4 else "::1"
PORT = 50007


def hook(event, args):
    print(f'\t[audit]: {event} with args={args}')


sys.addaudithook(hook)


# https://stackoverflow.com/a/77426018/330457
def randomize(a):
    a[:] = random.randbytes(len(a))


with closing(socket.socket(FAMILY, socket.SOCK_STREAM)) as s:
    # -------------------------------------------------------------------- bind
    if bool(random.getrandbits(1)):
        s.bind((HOST, 0))
        print(f'(optionally) bound to {s.getsockname()}')
    # ----------------------------------------------------------------- connect
    s.connect((HOST, PORT))
    print(f'connected to {s.getpeername()}')
    bytes_ = random.randint(0, 65536)
    print(f'sending {bytes_} byte(s)')
    array = bytearray(random.randint(1024, 8192))
    print(f'array.length: {len(array)}')
    hash_ = hashlib.sha256()
    while bytes_ > 0:
        # ---------------------------------------------------------------- send
        randomize(array)
        if bytes_ < len(array):
            array = array[0:bytes_]
        w = s.send(array)
        hash_.update(array)
        bytes_ -= w
        # ---------------------------------------------------------------- recv
        bytes_received = s.recv(len(array))
    # ---------------------------------------------------------------- shutdown
    s.shutdown(socket.SHUT_WR)
    # ------------------------------------------------------------- read-to-eof
    while True:
        bytes_received = s.recv(len(array))
        if len(bytes_received) == 0:
            break
    # ------------------------------------------------------------ print-digest
    print(f'digest: {b64encode(hash_.digest()).decode()}')

File: 04-verbose-hello-world-srv/00-verbose-hello-world-srv-basic/src/main/java/com/github/jinahya/hello/misc/c02rfc862/Rfc862Tcp1Client.java
package com.github.jinahya.hello.misc.c02rfc862;

/*-
 * #%L
 * verbose-hello-world-srv-common
 * %%
 * Copyright (C) 2018 - 2023 Jinahya, Inc.
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.github.jinahya.hello.util._ExcludeFromCoverage_PrivateConstructor_Obviously;
import lombok.extern.slf4j.Slf4j;

import java.io.EOFException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.concurrent.ThreadLocalRandom;

@Slf4j
@SuppressWarnings({"java:S127"})
class Rfc862Tcp1Client extends _Rfc862Tcp {

    public static void main(final String... args) throws Exception {
        try (var client = new Socket()) {
            // ---------------------------------------------------------------------- bind(optional)
            if (ThreadLocalRandom.current().nextBoolean()) {
                client.bind(new InetSocketAddress(HOST, 0));
                logBound(client);
            }
            // ----------------------------------------------------------------------------- connect
            client.connect(ADDR);
            logConnected(client);
            // ----------------------------------------------------------------------------- prepare
            final var digest = newDigest();
            final var array = newArray();
            // -------------------------------------------------------------------------- write/read
            for (var bytes = logClientBytes(newRandomBytes()); bytes > 0; ) {
                // --------------------------------------------------------------------------- write
                ThreadLocalRandom.current().nextBytes(array);
                final int l = Math.min(array.length, bytes);
                client.getOutputStream().write(array, 0, l);
                digest.update(array, 0, l);
                bytes -= l;
                // ---------------------------------------------------------------------------- read
                if (client.getInputStream().read(array) == -1) {
                    throw new EOFException("unexpected eof");
                }
            }
            // --------------------------------------------------------------- flush/shutdown-output
            client.getOutputStream().flush();
            client.shutdownOutput();
            // ----------------------------------------------------------- read-to-the-end-of-stream
            while (client.getInputStream().read(array) != -1) {
                // empty
            }
            // --------------------------------------------------------------------------------- log
            logDigest(digest);
        }
    }

    @_ExcludeFromCoverage_PrivateConstructor_Obviously
    private Rfc862Tcp1Client() {
        throw new AssertionError("instantiation is not allowed");
    }
}
